lib/system/arithm.nim

```diff
@@ -8,6 +8,75 @@
 #
 
 
+# Only clang has __has_builtin (so far)
+#
+# TODO: This is emitted at the wrong position so we don't actually have an
+#       emit. Could we add this to nimbase.h instead?
+{.emit: """#ifndef __has_builtin
+  #define __has_builtin(x) 0
+#endif""".}
+
+# Builtin compiler functions for improved performance
+
+proc checkFunction(name: string): string =
+  "((__has_builtin(__builtin_" & name & "_overflow)) || __GNUC__ >= 5)"
+
+# TODO: This is totally ugly. But we can't reliably detect this from Nim,
+# especially with cross-compiling where the user may be using an older compiler
+# version. Switching this on/off manually with a define seems weird as well.
```

Comment:
I think this should be done by a define. Once GCC 5.0 becomes widespread, we can enable it as a default in the config, for now it should be opt-in.